data_engineering.qmd

```diff
@@ -1,121 +1,526 @@
+   - Software developers might emphasize ease of integration.
+   - End-users would prioritize accuracy and responsiveness.
+
+5. **Understanding the Constraints and Limitations of Embedded Systems:**
+   Embedded devices come with their own set of challenges:
+   - Memory Limitations: KWS models need to be lightweight to fit within the memory constraints of embedded devices. Typically, KWS models might need to be as small as 16KB to fit in the always-on island of the SoC. Moreover, this is just the model size. Additional application code for pre-processing may also need to fit within the memory constraints.
+   - Processing Power: The computational capabilities of embedded devices are limited (few hundred MHz of clock speed), so the KWS model must be optimized for efficiency.
+   - Power Consumption: Since many embedded devices are battery-powered, the KWS system must be power-efficient.
+   - Environmental Challenges: Devices might be deployed in various environments, from quiet bedrooms to noisy industrial settings. The KWS system must be robust enough to function effectively across these scenarios.
+
+6. **Data Collection and Analysis:**
+   For a KWS system, the quality and diversity of data are paramount. Considerations might include:
+   - Variety of Accents: Collect data from speakers with various accents to ensure wide-ranging recognition.
+   - Background Noises: Include data samples with different ambient noises to train the model for real-world scenarios.
+   - Keyword Variations: People might either pronounce keywords differently or have slight variations in the wake word itself. Ensure the dataset captures these nuances.
+
+7. **Iterative Feedback and Refinement:**
+    Once a prototype KWS system is developed, it's crucial to test it in real-world scenarios, gather feedback, and iteratively refine the model. This ensures that the system remains aligned with the defined problem and objectives. This is important because the deployment scenarios change over time as things evolve.
 
-- Definition and Importance of Data Engineering in AI
-- Role of Data Engineering in Embedded AI
-- Synergy with Machine Learning and Deep Learning
 
-## Problem
 
-Explanation: This section is a crucial starting point in any data engineering project, as it lays the groundwork for the project's trajectory and ultimate success. Here's a brief explanation of why each subsection within the "Problem Definition" is important:
 
-- Identifying the Problem
-- Setting Clear Objectives
-- Benchmarks for Success
-- Stakeholder Engagement and Understanding
-- Understanding the Constraints and Limitations of Embedded Systems
-  
 ## Data Sourcing
+The quality and diversity of data gathered is important for developing accurate and robust AI systems. Sourcing high-quality training data requires careful consideration of the objectives, resources, and ethical implications. Data can be obtained from various sources depending on the needs of the project:
+
```

Comment:
This is generally really good, but it would be nice to see a section on ethics & data sourcing either at the beginning or end of this section. I think beginning would flow a bit better. You make brief references to ethics & bias throughout that would benefit from a more in depth section on why it matters.